tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15690,6 +15672,98 @@ void TryAddShaderAttrFromTargetProfile(Sema &S, FunctionDecl *FD,
   return;
 }
 
+// in the non-library case, this function will be run only once,
+// but in the library case, this function will be run for each
+// viable top-level function declaration by
+// ValidateNoRecursionInTranslationUnit.
+//  (viable as in, is exported)
+std::pair<clang::FunctionDecl *, bool>
+ValidateNoRecursion(clang::Sema *self, clang::FunctionDecl *FD) {
+  // Validate that there is no recursion; start with the entry function.
+  // NOTE: the information gathered here could be used to bypass code generation
+  // on functions that are unreachable (as an early form of dead code
+  // elimination).
+  FunctionDecl *pPatchFnDecl = nullptr;
+  if (FD) {
+    const auto *shaderModel =
+        hlsl::ShaderModel::GetByName(self->getLangOpts().HLSLProfile.c_str());
+
+    if (shaderModel->IsHS()) {
+      if (const HLSLPatchConstantFuncAttr *Attr =
+              FD->getAttr<HLSLPatchConstantFuncAttr>()) {
+        NameLookup NL = GetSingleFunctionDeclByName(
+            self, Attr->getFunctionName(), /*checkPatch*/ true);
+        if (!NL.Found || !NL.Found->hasBody()) {
+          unsigned id = self->Diags.getCustomDiagID(
+              clang::DiagnosticsEngine::Level::Error,
+              "missing patch function definition");
+          self->Diags.Report(id);
+          return std::make_pair(nullptr, true);
+        }
+        pPatchFnDecl = NL.Found;
+      }
+    }
+
+    hlsl::CallGraphWithRecurseGuard CG;
+    CG.BuildForEntry(FD);
+    FunctionDecl *pResult = CG.CheckRecursion(FD);
+    if (pResult) {
+      return std::make_pair(pResult, false);
+    }
+    if (pPatchFnDecl) {
+      CG.BuildForEntry(pPatchFnDecl);
+      pResult = CG.CheckRecursion(pPatchFnDecl);
+      if (pResult) {
+        return std::make_pair(pResult, true);
```

Comment:
Rather than test these one after the other, I think it makes sense to check for the missing patch constant function after the entry function by the caller. So instead of the above single call to ValidateNoRecursion and the complicated return value, we could test the entry function, if it's HS, test if there is a patch constant function and error if not, then test that for recursion with another call to this function.